lib/Sema/CSApply.cpp

```diff
@@ -4907,11 +4905,6 @@ namespace {
       // See whether there's an equivalent ObjC key path string we can produce
       // for interop purposes.
       checkAndSetObjCKeyPathString(E);
-      
-      // The final component type ought to line up with the leaf type of the
-      // key path.
-      assert(!componentTy || componentTy->hasUnresolvedType()
-             || componentTy->getWithoutSpecifierType()->isEqual(leafTy));
```

Comment:
Looks like the right check is probably `TypeChecker::isSubtypeOf` since that's the check we do for conversion of result types in `matchFunctionTypes`.